core/trino-main/src/main/java/io/trino/operator/TopNPeerGroupLookup.java

```diff
@@ -0,0 +1,374 @@
+    private long mask;
+
+    /**
+     * The hash strategy.
+     */
+    private final RowIdHashStrategy strategy;
+
+    /**
+     * The current allocated table size.
+     */
+    private long tableSize;
+
+    /**
+     * Threshold after which we rehash.
+     */
+    private long maxFill;
+
+    /**
+     * Number of entries in the set.
+     */
+    private long entryCount;
+
+    /**
+     * The acceptable load factor.
+     */
+    private final float fillFactor;
+
+    /**
+     * The value denoting unmapped group IDs. Since group IDs need to co-exist at all times with row IDs,
+     * we only need to use one of the two IDs to indicate that a slot is unused. Group IDs have been arbitrarily selected
+     * for that purpose.
+     */
+    private final long unmappedGroupId;
+
+    /**
+     * The default return value for {@code get()}, {@code put()} and {@code remove()}.
+     */
+    private final long defaultReturnValue;
+
+    /**
+     * Standard hash table parameters are expected. {@code unmappedGroupId} specifies the internal marker value for unmapped group IDs.
+     */
+    public TopNPeerGroupLookup(long expected, float fillFactor, RowIdHashStrategy strategy, long unmappedGroupId, long defaultReturnValue)
+    {
+        checkArgument(expected >= 0, "The expected number of elements must be nonnegative");
+        checkArgument(fillFactor > 0 && fillFactor <= 1, "Load factor must be greater than 0 and smaller than or equal to 1");
+        this.fillFactor = fillFactor;
+        this.strategy = requireNonNull(strategy, "strategy is null");
+        this.unmappedGroupId = unmappedGroupId;
+        this.defaultReturnValue = defaultReturnValue;
+
+        tableSize = bigArraySize(expected, fillFactor);
+        mask = tableSize - 1;
+        maxFill = maxFill(tableSize, fillFactor);
+        buffer = new Buffer(tableSize, unmappedGroupId);
+    }
+
+    public TopNPeerGroupLookup(long expected, RowIdHashStrategy strategy, long unmappedGroupId, long defaultReturnValue)
+    {
+        this(expected, 0.75f, strategy, unmappedGroupId, defaultReturnValue);
+    }
+
+    /**
+     * Returns the size of this hash map in bytes.
+     */
+    public long sizeOf()
+    {
+        return INSTANCE_SIZE + buffer.sizeOf();
+    }
+
+    public long size()
+    {
+        return entryCount;
+    }
+
+    public boolean isEmpty()
+    {
+        return entryCount == 0;
+    }
+
+    public long get(long groupId, long rowId)
+    {
+        checkArgument(groupId != unmappedGroupId, "Group ID cannot be the unmapped group ID");
+
+        long hash = hash(groupId, rowId);
+        long index = hash & mask;
+        if (buffer.isEmptySlot(index)) {
+            return defaultReturnValue;
+        }
+        if (hash == buffer.getPrecomputedHash(index) && equals(groupId, rowId, index)) {
+            return buffer.getValue(index);
+        }
+        // There's always an unused entry.
+        while (true) {
+            index = (index + 1) & mask;
```

Comment:
if you increment index at the end of loop you could get rid if `iffs` above

Reply:
yes, this could work. If you remember, this structure comes from the fastutil implementation style. I believe they were trying to unroll the loop to optimize for the common path. Not sure if actually makes a difference here though.

Reply:
Ok. That maybe makes sense